### DISCUSSION TITLE
Fluent Interface

### DISCUSSION THREAD

**Discussion by mihakralj** [DaveSkender/Stock.Indicators#796] [Dev notes]
Dave,
are you aware of Fluent API interface pattern as defined by Martin Fowler many moons ago?
https://dotnettutorials.net/lesson/fluent-interface-design-pattern/

I think this pattern fits perfectly to wrap your Stock.Indicators. Writing a super-class with methods that return a normalized object (that can be passed to the next method) would need no modifications of your library at all!

I implemented it in my dev branch and it looked really nice - before I abandoned the concept:
```
FluentClass strat = new();
strat.feed("AAPL",200).OHLC4.ZeroLag(5).DEMA(9).Slope(9).Above(1.1)
```

Why did I abandon it yet I recommend it to you? Well, I'm glad you asked... :-)
- fluent interface works with methods, not classes. 
- methods have no internal persistence that I need to prevent re-calculations
- chaining of methods is therefore ideal for static methods that operate on an object
- where did I see one huge class with tons of static methods? :-)

Seriously, I think you should give fluent interface a try and add a new fluent class with chainable normalized methods that can pass outputs to each other!

**Comment by DaveSkender** [comment]
I am definitely doing and going to do more method chaining, so I’ll be using this design pattern to some degree.  Some of the internal refactoring I’ve been doing lately will make this much easier.

It does not appear to favor to static classes, so it should still work with non-static classes, which will come in handy when we get around to streaming use cases.

The link you provided is interesting because it seems to be just a bunch of methods doing micro-changes on the same class which seems kind of silly to me, but maybe it’s meant to be on overly simple example.

Most of our [untilities](https://daveskender.github.io/Stock.Indicators/utilities/#content) use some form of method chaining using [extensions methods](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods).  It definitely makes the interface easier to use.

**Reply by mihakralj** [reply]
Better link for fluent interface pattern: https://en.m.wikipedia.org/wiki/Fluent_interface

**Reply by DaveSkender** [reply]
Interesting.  Yah, I suppose I am using this design pattern already to a large degree.  I didn’t know it had a name.  I see they refer to it as an ”API design pattern”, so that’s probably a good fit for a third-party library.

More from Martin Fowler: https://martinfowler.com/bliki/FluentInterface.html
